### PR TITLE
Support function getClientHTTPHeader

### PR DESCRIPTION
**Pull request by lingtaolf**
Resubmit: #54813

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add function `getClientHTTPHeader` for fetching values header values set in the HTTP request.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit bc6a8b079a1de2592601149266c5b3d7e3be9a21 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57665/bc6a8b079a1de2592601149266c5b3d7e3be9a21/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by lingtaolf** [comment]
@alexey-milovidov Please help review my code when you have time.

**Comment by alexey-milovidov** [comment]
Thank you! The linker shows duplicate symbols...

**Comment by alexey-milovidov** [comment]
Still duplicate.

File: tests/config/config.d/forbidden_get_client_http_headers.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<clickhouse>
+    <get_client_http_header_forbidden_headers>['FORBIDDEN-KEY1','FORBIDDEN-KEY2','FORBIDDEN-KEY,CONTAINS_COMMA']</get_client_http_header_forbidden_headers>
```
Comment (alexey-milovidov) [inline tests/config/config.d/forbidden_get_client_http_headers.xml:2]: XML is a structured language, and array formatting is another structured language. We shouldn't mix both.

File: src/Functions/getClientHTTPHeader.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        auto result_column = ColumnString::create();
+
+        const String default_value;
+        const std::unordered_set<String> & forbidden_header_list = getContext()->getClientHTTPHeaderForbiddenHeaders();
+
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            auto header_name = arg_string->getDataAt(row).toString();
+
+            if (!headers.has(header_name))
```
Comment (alexey-milovidov) [inline src/Functions/getClientHTTPHeader.cpp:92]: Better if it will return the default value.

File: src/Functions/getClientHTTPHeader.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "{} is not in HTTP request headers.", header_name);
+            else
+            {
+                auto it = forbidden_header_list.find(header_name);
```
Comment (alexey-milovidov) [inline src/Functions/getClientHTTPHeader.cpp:96]: contains?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
X-ClickHouse-Key should always be forbidden.

File: docs/en/sql-reference/functions/other-functions.md [hunk 4]
```diff
@@ -68,6 +68,45 @@ WHERE macro = 'test';
...
+
+
+When we use `clickhouse-client` to execute this function, we'll always get empty string, because client doesn't use http protocol.
+```sql
+SELECT getClientHTTPHeader('test')
+```
+result:  
+
+```text
+┌─getClientHTTpHeader('test')─┐
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/other-functions.md:97]: The documentation is wrong, should be HTTP, not HTTp.

**Comment by alexey-milovidov** [comment]
I'll continue and merge this PR in https://github.com/ClickHouse/ClickHouse/pull/61820.
Thank you! I will mention you as the author.

**Pull request closed without merging**